### PR TITLE
New create game flow backend - Part 1

### PR DESCRIPTION
**Pull request by TheFlash98**
### Contains
A start to the backend of the [new create game flow](https://goo.gl/hizreb). The `NewGameScreen` loads up the game templates along with their descriptions and the game name. The play button would load the game and select the defaultWorldGenerator for that world, is present else just select Perlin. The next screen loads up the `AdvanceGameSetupScreen` with all the modules downloaded and selected according to the game template choice. The next screen loads the available world generators associated with that game template and you can add it. Adding would create a new entry in the worlds dropdown. The new world is simply a String concatenated with a number incremented each time an entry is added.

### How to test
Play around with the setup as much as possible.

Video for what this PR covers: https://www.youtube.com/watch?v=G17jp6EHPqg

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/NewGameScreen.java [hunk 1]
```diff
@@ -15,24 +15,296 @@
...
+                    }
+                }
+            });
+        }
+        final UIText gameName = find("gameName",UIText.class);
+        setGameName(gameName);
+
+        final UIDropdownScrollable<Module> gameplay = find("gameplay", UIDropdownScrollable.class);
+        gameplay.setOptions(getGameplayModules());
+        gameplay.setVisibleOptions(3);
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/NewGameScreen.java]: With the UI a bit more spacious now could it perhaps make sense to bump this to 5?

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java [hunk 2]
```diff
@@ -0,0 +1,701 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java]: You might want to keep an eye out for the copyright year just as a good habit

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+public class AdvancedGameSetupScreen extends CoreScreenLayer {
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java]: Even if you're just moving a class you might want to add in some class-level javadoc - by moving it you indicate you might know a thing about what it does and now you touched it last -  tag you're it! :-)

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+
+    private static final Logger logger = LoggerFactory.getLogger(SelectModulesScreen.class);
+    private final Comparator<? super ModuleSelectionInfo> moduleInfoComparator = (o1, o2) -> o1.getMetadata()
+            .getDisplayName().toString().compareTo(o2.getMetadata().getDisplayName().toString());
+    @In
+    private ModuleManager moduleManager;
+    @In
+    private Config config;
+
+    private SelectModulesConfig selectModulesConfig;
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/selectModulesScreen/AdvancedGameSetupScreen.java]: This looks slightly out of order without an `@In` yet in the midst of other `@In`ed thing. Could move it down to the private grouping

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by TheFlash98** [comment]
[theflash98 & multi-world](https://trello.com/c/dYU1iOul/38-theflash98-multi-world)

**Pull request merged**